File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/ProviderAwareRewriteScore.java
package com.foundgine.core.semantic.planning;

import java.util.Objects;

/** Combined deterministic score for a rewrite under a provider cost model. */
public record ProviderAwareRewriteScore(double value) {
	public static ProviderAwareRewriteScore calculate(RewriteBenefit benefit, RewriteCost rewriteCost,
			ProviderCostEstimate providerCost, ProviderCostSelectionPolicy policy) {
		Objects.requireNonNull(policy);
		policy.validate();
		double providerPenalty = policy.preferLowerProviderCost()
				? providerCost.estimatedExecutionCost() * policy.providerCostWeight()
				: 0d;
		return new ProviderAwareRewriteScore(
				benefit.estimatedBenefit() / (1d + rewriteCost.estimatedWork() + providerPenalty));
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/ProviderAwareRewriteScore.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Combined deterministic score for a rewrite under a specific provider cost model.</summary>
public readonly record struct ProviderAwareRewriteScore(double Value)
{
    public static ProviderAwareRewriteScore Calculate(
        RewriteBenefit benefit,
        RewriteCost rewriteCost,
        ProviderCostEstimate providerCost,
        ProviderCostSelectionPolicy policy)
    {
        ArgumentNullException.ThrowIfNull(policy);
        policy.Validate();

        var rewritePenalty = rewriteCost.EstimatedWork;
        var providerPenalty = policy.PreferLowerProviderCost
            ? providerCost.EstimatedExecutionCost * policy.ProviderCostWeight
            : 0d;
        var denominator = 1d + rewritePenalty + providerPenalty;
        var score = benefit.EstimatedBenefit / denominator;
        return new ProviderAwareRewriteScore(score);
    }
}
